### PR TITLE
Reimplement #5339 by justbur to add :packages keyword.

### PR DESCRIPTION
**Pull request by ralesi**
Adding the changes proposed in #5339 by @justbur to develop.  Has no issues, and the original bug with `nil` in `helm-spacemacs-help` has been resolved by the recent layer changes by @syl20bnr.

### CODE REVIEW

**Comment by ralesi** [comment]
To restrict layer to only use pkg1 and pkg2 use
```
(layer
:packages (pkg1 pkg2)
```
To load all packages in a layer except pkg1 use
```
(layer
:packages (not pkg1))
```

**Comment by TheBB** [comment]
Except for some strange indentation, this looks clean and sensible to me.

**Comment by NJBS** [comment]
Can this be documented somewhere as well?

**Comment by ralesi** [comment]
@NJBS, updated documentation and rebased to current HEAD

**Comment by syl20bnr** [comment]
Thank you ! 💜 
I reimplemented it to improve the overall code for `configuration-layer/get-packages` which became hard to follow (well it is still a complicated function). Also the reimplementation does not use the `:excluded` slot of package objects which was important to me, so now not selected packages are local to the layer where they are not selected, it allows more granular control on what the user wants to disable. To completely disable a package cross-platform there is the `dotspacemas-excluded-packages` variable.
More info in the horrible commit message here https://github.com/syl20bnr/spacemacs/commit/57350a3 (sorry for this).
Cherry-picked into develop branch, you can safely delete your branch.
